File: agent-service/src/agents/testing/test-runner.ts
/**
 * Test runner — detects the test framework for a workspace and executes tests,
 * capturing stdout/stderr and parsing pass/fail counts.
 */
import { exec } from "child_process";
import { promisify } from "util";
import * as fs from "fs/promises";
import * as path from "path";

const execAsync = promisify(exec);

export interface TestFramework {
  name: string;
  command: string;
  /** Regex patterns to extract pass/fail counts from output */
  passPattern: RegExp;
  failPattern: RegExp;
}

const KNOWN_FRAMEWORKS: TestFramework[] = [
  {
    name: "vitest",
    command: "npx vitest run --reporter verbose 2>&1",
    passPattern: /(\d+)\s+passed/i,
    failPattern: /(\d+)\s+failed/i,
  },
  {
    name: "jest",
    command: "npx jest --no-coverage 2>&1",
    passPattern: /Tests:\s+.*?(\d+)\s+passed/i,
    failPattern: /Tests:\s+.*?(\d+)\s+failed/i,
  },
  {
    name: "mocha",
    command: "npx mocha 2>&1",
    passPattern: /(\d+)\s+passing/i,
    failPattern: /(\d+)\s+failing/i,
  },
  {
    name: "pytest",
    command: "python -m pytest -v 2>&1",
    passPattern: /(\d+)\s+passed/i,
    failPattern: /(\d+)\s+failed/i,
  },
  {
    name: "go-test",
    command: "go test ./... 2>&1",
    passPattern: /ok\s+/i,
    failPattern: /FAIL\s+/i,
  },
  {
    name: "cargo-test",
    command: "cargo test 2>&1",
    passPattern: /(\d+)\s+passed/i,
    failPattern: /(\d+)\s+failed/i,
  },
];

export interface TestRunResult {
  framework: string;
  passed: number;
  failed: number;
  output: string;
  success: boolean;
  /** Duration in milliseconds */
  durationMs: number;
}

/**
 * Detect which test framework a workspace uses by inspecting package.json,
 * Cargo.toml, go.mod, requirements.txt, etc.
 */
export async function detectFramework(workspacePath: string): Promise<TestFramework | null> {
  // Check package.json for JS/TS projects
  try {
    const pkgRaw = await fs.readFile(path.join(workspacePath, "package.json"), "utf-8");
    const pkg = JSON.parse(pkgRaw) as {
      scripts?: Record<string, string>;
      devDependencies?: Record<string, string>;
      dependencies?: Record<string, string>;
    };

    const allDeps = {
      ...pkg.devDependencies,
      ...pkg.dependencies,
    };

    if (allDeps.vitest || pkg.scripts?.test?.includes("vitest")) {
      return KNOWN_FRAMEWORKS.find((f) => f.name === "vitest") ?? null;
    }
    if (allDeps.jest || pkg.scripts?.test?.includes("jest")) {
      return KNOWN_FRAMEWORKS.find((f) => f.name === "jest") ?? null;
    }
    if (allDeps.mocha || pkg.scripts?.test?.includes("mocha")) {
      return KNOWN_FRAMEWORKS.find((f) => f.name === "mocha") ?? null;
    }
  } catch {
    // Not a JS/TS project
  }

  // Check for Python
  try {
    await fs.access(path.join(workspacePath, "pytest.ini"));
    return KNOWN_FRAMEWORKS.find((f) => f.name === "pytest") ?? null;
  } catch { /* continue */ }

  try {
    const req = await fs.readFile(path.join(workspacePath, "requirements.txt"), "utf-8");
    if (req.includes("pytest")) return KNOWN_FRAMEWORKS.find((f) => f.name === "pytest") ?? null;
  } catch { /* continue */ }

  // Check for Go
  try {
    await fs.access(path.join(workspacePath, "go.mod"));
    return KNOWN_FRAMEWORKS.find((f) => f.name === "go-test") ?? null;
  } catch { /* continue */ }

  // Check for Rust
  try {
    await fs.access(path.join(workspacePath, "Cargo.toml"));
    return KNOWN_FRAMEWORKS.find((f) => f.name === "cargo-test") ?? null;
  } catch { /* continue */ }

  return null;
}

/**
 * Run the detected test framework in the given workspace directory.
 * Returns a structured TestRunResult regardless of pass/fail.
 */
export async function runTests(
  workspacePath: string,
  framework: TestFramework,
  timeoutMs = 300_000, // 5 minute default
): Promise<TestRunResult> {
  const start = Date.now();

  let output = "";
  let exitCode = 0;

  try {
    const { stdout, stderr } = await execAsync(framework.command, {
      cwd: workspacePath,
      timeout: timeoutMs,
      maxBuffer: 5 * 1024 * 1024,
    });
    output = stdout + stderr;
  } catch (err: unknown) {
    const execErr = err as { stdout?: string; stderr?: string; code?: number };
    output = (execErr.stdout ?? "") + (execErr.stderr ?? "");
    exitCode = execErr.code ?? 1;
  }

  const durationMs = Date.now() - start;

  const passMatch = framework.passPattern.exec(output);
  const failMatch = framework.failPattern.exec(output);

  const passed = passMatch ? parseInt(passMatch[1] ?? "0", 10) : (exitCode === 0 ? 1 : 0);
  const failed = failMatch ? parseInt(failMatch[1] ?? "0", 10) : (exitCode !== 0 ? 1 : 0);

  return {
    framework: framework.name,
    passed,
    failed,
    output: output.slice(0, 8000), // cap output for storage
    success: exitCode === 0 && failed === 0,
    durationMs,
  };
}

File: agent-service/src/agent_service/agents/testing/test_runner.py
from __future__ import annotations

import asyncio
import json
import os
from dataclasses import dataclass
from typing import List, Optional


@dataclass
class TestFramework:
    name: str
    command: str
    pass_pattern: str
    fail_pattern: str


@dataclass
class TestRunResult:
    framework: str
    passed: int
    failed: int
    output: str
    success: bool
    duration_ms: int


_KNOWN_FRAMEWORKS = [
    TestFramework("vitest", "npx vitest run --reporter verbose 2>&1", r"(\d+)\s+passed", r"(\d+)\s+failed"),
    TestFramework("jest", "npx jest --no-coverage 2>&1", r"Tests:\s+.*?(\d+)\s+passed", r"Tests:\s+.*?(\d+)\s+failed"),
    TestFramework("mocha", "npx mocha 2>&1", r"(\d+)\s+passing", r"(\d+)\s+failing"),
    TestFramework("pytest", "python -m pytest -v 2>&1", r"(\d+)\s+passed", r"(\d+)\s+failed"),
    TestFramework("go-test", "go test ./... 2>&1", r"ok\s+", r"FAIL\s+"),
    TestFramework("cargo-test", "cargo test 2>&1", r"(\d+)\s+passed", r"(\d+)\s+failed"),
]


async def detect_framework(workspace_path: str) -> Optional[TestFramework]:
    """Detect the test framework used by the workspace."""
    pkg_json = os.path.join(workspace_path, "package.json")
    if os.path.isfile(pkg_json):
        try:
            with open(pkg_json, "r", encoding="utf-8") as fh:
                pkg = json.load(fh)
            all_deps = {**pkg.get("devDependencies", {}), **pkg.get("dependencies", {})}
            scripts = pkg.get("scripts", {})
            if "vitest" in all_deps or "vitest" in scripts.get("test", ""):
                return next(f for f in _KNOWN_FRAMEWORKS if f.name == "vitest")
            if "jest" in all_deps or "jest" in scripts.get("test", ""):
                return next(f for f in _KNOWN_FRAMEWORKS if f.name == "jest")
            if "mocha" in all_deps or "mocha" in scripts.get("test", ""):
                return next(f for f in _KNOWN_FRAMEWORKS if f.name == "mocha")
        except Exception:
            pass

    for indicator, name in [
        ("pytest.ini", "pytest"), ("pyproject.toml", "pytest"), ("requirements.txt", "pytest"),
        ("go.mod", "go-test"), ("Cargo.toml", "cargo-test"),
    ]:
        indicator_path = os.path.join(workspace_path, indicator)
        if os.path.isfile(indicator_path):
            if name == "pytest":
                try:
                    with open(indicator_path, "r") as fh:
                        if "pytest" in fh.read() or indicator == "pytest.ini":
                            return next(f for f in _KNOWN_FRAMEWORKS if f.name == "pytest")
                    continue
                except Exception:
                    pass
            return next(f for f in _KNOWN_FRAMEWORKS if f.name == name)

    return None


async def run_tests(
    workspace_path: str,
    framework: TestFramework,
    timeout_ms: int = 300_000,
) -> TestRunResult:
    """Run the detected test framework in the workspace directory."""
    import re
    import time

    start = time.monotonic()
    output = ""
    exit_code = 0

    try:
        proc = await asyncio.create_subprocess_shell(
            framework.command,
            cwd=workspace_path,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.STDOUT,
        )
        try:
            stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=timeout_ms / 1000)
            output = stdout.decode(errors="replace")
            exit_code = proc.returncode or 0
        except asyncio.TimeoutError:
            proc.kill()
            output = "Tests timed out"
            exit_code = 1
    except Exception as exc:
        output = str(exc)
        exit_code = 1

    duration_ms = int((time.monotonic() - start) * 1000)

    pass_match = re.search(framework.pass_pattern, output, re.IGNORECASE)
    fail_match = re.search(framework.fail_pattern, output, re.IGNORECASE)

    passed = int(pass_match.group(1)) if pass_match and pass_match.lastindex else (1 if exit_code == 0 else 0)
    failed = int(fail_match.group(1)) if fail_match and fail_match.lastindex else (1 if exit_code != 0 else 0)

    return TestRunResult(
        framework=framework.name,
        passed=passed,
        failed=failed,
        output=output[:8000],
        success=exit_code == 0 and failed == 0,
        duration_ms=duration_ms,
    )
